pop_tools/budget.py

```diff
@@ -0,0 +1,328 @@
+    ladv = (ladv_zonal + ladv_merid).rename('ladv').sum('z_t')
+    ladv = _convert_units(ladv)
+    ladv.attrs['long_name'] = 'lateral advection'
+    return ladv.load()
+
+
+def _compute_lateral_mixing(ds, grid, mask, kmax=None):
+    """Compute lateral mixing component."""
+    print('Computing lateral mixing...')
+    ds = ds[['HDIFN', 'HDIFE', 'HDIFB']]
+    if kmax is not None:
+        ds = ds.isel(z_t=slice(0, kmax + 1))
+    ds = ds.groupby('time.year').mean('time').rename({'year': 'time'})
+    # Flip sign so that positive direction is upwards.
+    ds = _convert_to_tendency(ds, grid.vol, sign=-1, kmax=kmax)
+    lmix_zonal = _compute_horizontal_divergence(ds.HDIFE, mask, direction='zonal')
+    lmix_merid = _compute_horizontal_divergence(ds.HDIFN, mask, direction='meridional')
+    lmix_B = ds.HDIFB.where(mask)
+    lmix_B = _compute_vertical_divergence(lmix_B)
+
+    # Sum all lateral mixing components
+    lmix = (lmix_merid + lmix_zonal + lmix_B).rename('lmix').sum('z_t')
+    lmix = _convert_units(lmix)
+    lmix.attrs['long_name'] = 'lateral mixing'
+    return lmix.load()
+
+
+def _compute_vertical_advection(ds, grid, mask, kmax=None):
+    """Compute vertical advection (WT)"""
+    print('Computing vertical advection...')
+    ds = ds['WT']
+    if kmax is not None:
+        # Need one level below max depth to compute divergence into bottom layer.
+        ds = ds.isel(z_t=slice(0, kmax + 2))
+    ds = ds.groupby('time.year').mean('time').rename({'year': 'time'})
+    ds = ds.where(mask)
+    if kmax is not None:
+        ds = _convert_to_tendency(ds, grid.vol, kmax=kmax + 1)
+    else:
+        ds = _convert_to_tendency(ds, grid.vol)
+
+    # Compute divergence of vertical advection.
+    vadv = (ds.shift(z_t=-1).fillna(0) - ds).isel(z_t=slice(0, -1))
+    vadv = vadv.sum('z_t').rename('vadv')
+    vadv = _convert_units(vadv)
+    vadv.attrs['long_name'] = 'vertical advection'
+    return vadv.load()
+
+
+def _compute_vertical_mixing(ds, grid, mask, kmax=None):
+    """Compute contribution from vertical mixing."""
+    print('Computing vertical mixing...')
+    ds = ds[['DIA_IMPVF', 'KPP_SRC']]
+    if kmax is not None:
+        ds = ds.isel(z_t=slice(0, kmax + 1))
+    ds = ds.groupby('time.year').mean('time').rename({'year': 'time'})
+    ds = ds.where(mask)
+    # Only need to flip sign of DIA_IMPVF.
+    ds['DIA_IMPVF'] = _convert_to_tendency(ds['DIA_IMPVF'], grid.area, sign=-1, kmax=kmax)
+    ds['KPP_SRC'] = _convert_to_tendency(ds['KPP_SRC'], grid.vol, kmax=kmax)
+
+    # Compute divergence of diapycnal mixing
+    diadiff = _compute_vertical_divergence(ds.DIA_IMPVF)
+
+    # Sum KPP and diapycnal mixing to create vmix term.
+    vmix = (ds.KPP_SRC + diadiff).rename('vmix').sum('z_t')
+    vmix = _convert_units(vmix)
+    vmix.attrs['long_name'] = 'vertical mixing'
+    return vmix.load()
+
+
+def _compute_SMS(ds, grid, mask, kmax=None):
+    """Compute SMS term from biology."""
+    print('Computing source/sink...')
```

Comment:
Since we are dealing with `dask`'s lazy computation, wouldn't it be confusing to some users having this line printed when in reality the computation hasn't been triggered yet?

Reply:
That makes sense. So we could either move this to just before `ds.load()`, or could do away with it completely. I think there should be *some* level of feedback here since this code takes a decent amount of time to execute (~5 minutes with all terms currently). We could do print statements, or use `progressbar`. Or just assume that users are using `dask` to check progression.